### PR TITLE
[sw,crypto] Update crypto API header file to align with the latest crypto library spec

### PR DESCRIPTION
**Pull request by y-srini**
The crypto API header file is updated to be in sync with the latest crypto lib specification changes (as of 16Jun22)

Signed-off-by: Srinivasan Y <srinivasan.yadhunathan@seagate.com>

### CODE REVIEW

File: sw/device/lib/crypto/api.h [hunk 1]
```diff
@@ -259,1090 +445,1385 @@ typedef enum eddsa_sign_mode {
...
+crypto_status_t otcrypto_mac(const crypto_blinded_key_t *key,
+                             crypto_const_uint8_buf_t input_message,
+                             mac_mode_t mac_mode,
+                             crypto_uint8_buf_t customization_string,
+                             size_t required_output_len,
+                             crypto_uint8_buf_t *digest);
 
 /**
- * Generic HMAC context.
+ * Performs the INIT operation for a HMAC function.
```
Comment (jadephilipoom) [inline sw/device/lib/crypto/api.h:1018]:
```suggestion
 * Performs the INIT operation for HMAC.
```

File: sw/device/lib/crypto/api.h [hunk 2]
```diff
@@ -259,1090 +445,1385 @@ typedef enum eddsa_sign_mode {
...
+ * @return crypto_status_t Result of the HMAC update operation
  */
-crypto_status_t hmac_sha256_init(const hash_context_t *ctx,
-                                 const crypto_blinded_key_t *key);
+crypto_status_t otcrypto_hmac_update(hmac_context_t *const ctx,
+                                     crypto_const_uint8_buf_t input_message);
 
 /**
- * Performs the hmac-SHA2-256 update operation.
+ * Performs the generic HMAC FINAL operation.
```
Comment (jadephilipoom) [inline sw/device/lib/crypto/api.h:1060]:
```suggestion
 * Performs the FINAL operation for HMAC.
```

File: sw/device/lib/crypto/api.h [hunk 3]
```diff
@@ -259,1090 +445,1385 @@ typedef enum eddsa_sign_mode {
...
+ * operation.
  */
-typedef struct drbg_state drbg_state_t;
+crypto_status_t otcrypto_x25519_keygen_async_start(
+    drbg_state_t *drbg_state, crypto_uint8_buf_t additional_input,
+    size_t required_key_len);
 
 /**
- * Initializes the DRBG.
+ * Finalizes the asynchronous key generation for X25519 operation.
```
Comment (jadephilipoom) [inline sw/device/lib/crypto/api.h:1736]:
```suggestion
 * Finalizes the asynchronous key generation for X25519.
```

File: sw/device/lib/crypto/api.h [hunk 4]
```diff
@@ -22,214 +22,398 @@ extern "C" {
  * Values are hardened.
  */
 typedef enum crypto_status {
-  // Status is OK; no errors.
-  kCryptoStatusOK = 0x7d3a,
+  // Status is OK; No errors.
```
[the remark this answers is not in the document]
Comment (y-srini) [reply sw/device/lib/crypto/api.h:25]: resolved.

File: sw/device/lib/crypto/api.h [hunk 5]
```diff
@@ -22,214 +22,398 @@ extern "C" {
...
  * Values are hardened.
  */
-typedef enum verification_status {
-  // Return value for successful verification.
-  kVerificationStatusPass = 0x2c89,
-  // Return value for unsuccessful verification.
-  kVerificationStatusFail = 0x49d3,
-} verification_status_t;
+typedef enum key_type {
+  // Key type AES
```
[the remark this answers is not in the document]
Comment (y-srini) [reply sw/device/lib/crypto/api.h:41]: Added full-stop ( . ) at the end, for the comments that were missing them.

File: sw/device/lib/crypto/api.h [hunk 6]
```diff
@@ -259,1090 +445,1385 @@ typedef enum eddsa_sign_mode {
...
+ * Enum to define the supported DRBG mechanisms.
  *
- * @param key Pointer to the blinded gcm-key struct
- * @param ciphertext Input data to be decrypted
- * @param iv Initialization vector for the decryption function
- * @param aad Additional authenticated data
- * @param auth_tag Authentication tag to be verified
- * @param plaintext Decrypted plaintext data, same len as input data
- * @return crypto_status_t The result of the encryption operation
+ * Values are hardened.
```
[the remark this answers is not in the document]
Comment (y-srini) [reply sw/device/lib/crypto/api.h:506]: The values are hardened now.

File: sw/device/lib/crypto/api.h [hunk 7]
```diff
@@ -22,214 +22,398 @@ extern "C" {
...
 
 /**
- * Type of blinded keys.
+ * Enum to denote the AES-GCM tag length.
  *
- * This type is not exposed in the API because it should not be used directly.
+ * Values are hardened.
  */
-typedef struct crypto_blinded_key crypto_blinded_key_t;
+typedef enum aead_gcm_tag_len {
```
[the remark this answers is not in the document]
Comment (y-srini) [reply sw/device/lib/crypto/api.h:191]: Added a comment to the doc to discuss. Will leave it as such for now, and will update the enum with finalized values later.

File: sw/device/lib/crypto/api.h [hunk 8]
```diff
@@ -259,1090 +445,1385 @@ typedef enum eddsa_sign_mode {
...
+ * Performs the SHA2-256 FINAL operation.
+ *
+ * The final function process the remaining data and copies the final
+ * hash to the `digest` parameter.
  *
- * The final operation computes the SHA2-256 hash on the remaining
- * partial blocks if any, and then computes the final hash and stores
- * it in the `digest` parameter.
+ * This API is internal to the generic hash API and should not be
+ * called directly.
```
[the remark this answers is not in the document]
Comment (y-srini) [reply sw/device/lib/crypto/api.h:882]: Yup, was thinking about it too. Don't think we need to have these internal functions in the API header. Will remove them.

**Review by jadephilipoom — APPROVED** [review]
Looks good to me!

**Pull request merged**